src/widgets/auto-styler.js

```diff
@@ -0,0 +1,141 @@
+var AutoStyler = cdb.core.Model.extend({
+  initialize: function (dataviewModel) {
+    this.dataviewModel = dataviewModel;
+    this.colors = new CategoryColors();
+    this.vector = !!this.dataviewModel._dataProvider;
+  },
+
+  getStyle: function () {
+    var style;
+    var widgetType = this.dataviewModel.get('type');
+    if (widgetType === 'category') {
+      style = this._getStyleForCategory();
+    } else if (widgetType === 'histogram') {
+      style = this._getStyleForHistogram();
+    }
+    return style;
+  },
+
+  _getStyleForHistogram: function () {
+    var style = '';
+    var colors = ['YlGnBu', 'Greens', 'Reds', 'Blues'];
+    var color = colors[Math.floor(Math.random() * colors.length)];
+    var stylesByGeometry = AutoStyler.STYLE_TEMPLATE;
+    if (this.vector) {
+      var geometryType = this._getGeometryType();
+      style = this._getHistGeometry(geometryType)
+        .replace('{{layername}}', '#layer{');
+    } else {
+      for (var symbol in stylesByGeometry) {
+        style += this._getHistGeometry(symbol)
+          .replace('{{layername}}', this._getLayerHeader(symbol));
+      }
+    }
+    return style.replace(/{{column}}/g, this.dataviewModel.get('column'))
+      .replace(/{{bins}}/g, this.dataviewModel.get('bins'))
+      .replace(/{{color}}/g, color)
+      .replace(/{{min}}/g, 1)
+      .replace(/{{max}}/g, 20)
+      .replace(/{{ramp}}/g, '')
+      .replace(/{{defaultColor}}/g, '#000');
+  },
+
+  _getHistGeometry: function (geometryType) {
+    var style = AutoStyler.STYLE_TEMPLATE[geometryType];
+    if (geometryType === 'polygon') {
+      style = style.replace('{{defaultColor}}', 'ramp([{{column}}], colorbrewer({{color}}, {{bins}}))');
+    } else if (geometryType === 'marker') {
+      style = style.replace('{{markerWidth}}', 'ramp([{{column}}], {{min}}, {{max}}), {{bins}})');
+    } else {
+      style = style.replace('{{defaultColor}}', '#000');
+    }
+    return style;
+  },
+
+  _getStyleForCategory: function () {
+    var style = '';
+    var defColor = this.colors.getColorByCategory('Other');
+    var stylesByGeometry = AutoStyler.STYLE_TEMPLATE;
+    if (this.vector) {
+      var geometryType = this._getGeometryType();
+      var ramp = this._generateCategoryRamp(geometryType);
+      style = stylesByGeometry[geometryType]
+        .replace('{{ramp}}', ramp)
+        .replace('{{layername}}', '#layer{');
+    } else {
+      for (var symbol in stylesByGeometry) {
+        style += stylesByGeometry[symbol]
+          .replace('{{layername}}', this._getLayerHeader(symbol))
+          .replace('{{ramp}}', this._generateCategoryRamp(symbol));
+      }
+    }
+    return style
+      .replace(/{{defaultColor}}/g, defColor)
+      .replace('{{markerWidth}}', 10);
+  },
+
+  _getLayerHeader: function (symbol) {
+    return '#' + this.dataviewModel.layer.get('layer_name') + '[mapnik-geometry-type=' + AutoStyler.MAPNIK_MAPPING[symbol] + ']{';
+  },
+
+  _getGeometryType: function () {
```

Comment:
For example. if deep-insights.js needs to get the geometry type from a layer, we can implement this in CartoDB.js so that we can just do something like `this.dataviewModel.layer.getGeometryType()` here (and remove the `if (this.vector) { ... } else { ... } `...